### ISSUE TITLE
{error,emfile}

### ISSUE THREAD

**Issue by skvamme** [ninenines/cowboy#1031]
Getting error in Cowboy. Is this something you have seen before or should I investigate?
Setup is Cowboy and Gun connected over several days and lots of websocket messages.

23:12:05.145 [error] CRASH REPORT Process <0.31357.2> with 0 neighbours exited with reason: no match of right hand value {error,emfile} in filename:absname/1 line 69 in cowboy_stream_h:proc_lib_hack/3 line 162

**Comment by essen** [comment]
For some reason your node fails to return a proper `file:get_cwd()`. See https://github.com/erlang/otp/blob/maint/lib/stdlib/src/filename.erl#L69

Sounds like a problem on your end. Do investigate.

**Comment by skvamme** [comment]
It has probably nothing to do with websocket messages. Problem seems to appear after calling static html-page 10000nds of times. I'll try to dig deeper.

2016-10-11 10:36:20 =CRASH REPORT====
  crasher:
    initial call: cowboy_stream_h:proc_lib_hack/3
    pid: <0.14051.68>
    registered_name: []
    exception exit: {{{badmatch,{error,emfile}},[{filename,absname,1,[{file,"filename.erl"},{line,69}]},{cowboy_static,init_dir,3,[{file,"src/cowboy_static.erl"},{line,82}]},{cowboy_handler,execute,2,[{file,"src/cowboy_handler.erl"},{line,39}]},{cowboy_stream_h,execute,3,[{file,"src/cowboy_stream_h.erl"},{line,172}]},{cowboy_stream_h,proc_lib_hack,3,[{file,"src/cowboy_stream_h.erl"},{line,157}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]},[{cowboy_stream_h,proc_lib_hack,3,[{file,"src/cowboy_stream_h.erl"},{line,162}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [<0.13983.68>,<0.395.0>,<0.394.0>,ranch_sup,<0.377.0>]
    messages: []
    links: [<0.13983.68>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 987
    stack_size: 27
    reductions: 299
  neighbours:

**Comment by essen** [comment]
Curious. The process where cowboy_static executes gets deleted after the response is sent and resources are supposed to be freed.

**Comment by skvamme** [comment]
After running my test suite for three days I got the following error, from Gun this time
=ERROR REPORT==== 30-Oct-2016::21:07:04 ===
Error in process <0.9234.45> on node 'shooter@127.0.0.1' with exit value:
{{badmatch,{error,{failed_connect,[{to_address,{"127.0.0.1",8080}},
                                   {inet,[inet],emfile}]}}},
 [{comet_sensor_poll,loop,3,[{file,"src/comet_sensor_poll.erl"},{line,30}]}]}

comet_sensor_poll line 30 is
{ok,{_,_,Data}} = httpc:request(get,{?URL,[]},[],[{body_format,binary}]),


Restarting Cowboy did not clear the error, but restarting Gun did!

**Comment by essen** [comment]
So it's an issue in Gun then. :-)

**Comment by essen** [comment]
Wondering if you could check if there's a lot of processes in Gun's supervision tree perhaps.

**Comment by essen** [comment]
Closing because apparently not a Cowboy issue. Please open a new ticket in Gun with more details (maybe a supervision tree or something), or if it turns out to be Cowboy at fault, reopen this ticket. Thanks!

**Comment by skvamme** [comment]
Yes, a bit hard to find out what it is, but right now it seems to be an OS issue with ephemeral port shortage.

**Comment by essen** [comment]
Reopening because I would like to avoid using filename:absname/1 if we can have an alternative that doesn't require using a fd.
